### ISSUE TITLE
OutOfMemoryException or StackOverflowException while generating error message

### ISSUE THREAD

**Issue by dangerozov** [derekgreer/expectedObjects#38]
When ExpectedObjects finds that objects aren't equal it tries to serialize them into a string. Depending on the structure of the object this could lead to OutOfMemoryException or StackOverflow exception.

Example:
```
        [Test]
        public void Try()
        {
            new Another { ID = 1, Inners = new[] { new Inner { ID = 1} } }.ToExpectedObject().ShouldMatch(new Another { ID = 10, Inners = new[] { new Inner { ID = 10 } } });
        }

        public class Another
        {
            private int _ID;
            public int ID { get { return _ID; } set { _ID = value; } }

            public Inner[] Inners { get; set; }
        }

        public class Inner
        {
            public int ID { get; set; }
            public EnumEmu Type => EnumEmu.One;
        }

        public class EnumEmu
        {
            public int ID { get; set; }

            public static EnumEmu One => new EnumEmu { ID = 1 };
            public static EnumEmu Two => new EnumEmu { ID = 2 };
            public static EnumEmu Three => new EnumEmu { ID = 3 };

            public override bool Equals(object obj)
            {
                return obj is EnumEmu value
                    ? ID == value.ID
                    : false;
            }

            public override int GetHashCode()
            {
                return ID.GetHashCode();
            }
        }
```

This code produces following error message.

```
    ExpectedObjects.ComparisonException : The expected object did not match the actual object.
    
    The following issues were found:
    
    1) Another:
    
      Expected:
        Another
        { 
            ID = 10,
            Inners = new Inner[] { new Inner
            { 
                ID = 10,
                Type = new EnumEmu
                { 
                    ID = 1,
                    One = ... ,
                    Two = new EnumEmu
                    { 
                        ID = 2,
                        One = ... ,
                        Two = ... ,
                        Three = new EnumEmu
                        { 
                            ID = 3,
                            One = ... ,
                            Two = ... ,
                            Three = ... 
                        }
                    },
                    Three = new EnumEmu
                    { 
                        ID = 3,
                        One = ... ,
                        Two = new EnumEmu
                        { 
                            ID = 2,
                            One = ... ,
                            Two = ... ,
                            Three = ... 
                        },
                        Three = ... 
                    }
                }
            }}
        }
    
      Actual:
        Another
        { 
            ID = 1,
            Inners = new Inner[] { new Inner
            { 
                ID = 1,
                Type = new EnumEmu
                { 
                    ID = 1,
                    One = ... ,
                    Two = new EnumEmu
                    { 
                        ID = 2,
                        One = ... ,
                        Two = ... ,
                        Three = new EnumEmu
                        { 
                            ID = 3,
                            One = ... ,
                            Two = ... ,
                            Three = ... 
                        }
                    },
                    Three = new EnumEmu
                    { 
                        ID = 3,
                        One = ... ,
                        Two = new EnumEmu
                        { 
                            ID = 2,
                            One = ... ,
                            Two = ... ,
                            Three = ... 
                        },
                        Three = ... 
                    }
                }
            }}
        }
```

If I replace `public Inner[] Inners { get; set; }` with `public EnumEmu Type => EnumEmu.One;` I get following message

```
ExpectedObjects.ComparisonException : The expected object did not match the actual object.
    
    The following issues were found:
    
    1) Another.ID:
    
      Expected:
        10
    
      Actual:
        1
```

But if I remove override of Equals/GetHashCode, I get StackOverflowException.

If I add more static properties to `EnumEmu`, like 40 more, I get OutOfMemoryException.

**Comment by dangerozov** [comment]
And why does it serialize static properties? They're not even part of the object.

**Comment by derekgreer** [comment]
This issue of displaying static members in exception message output has been resolved in version 3.5.3.

A note on the problem presented:

While the display of static members in the exception message was an issue, the underlying issue in the provided example is one of attempting to compare objects whose properties are both recursive (i.e. types with member hierarchies which reference the same type) and where each instance is semantically and ontologically unique (i.e. MemberInstance != MemberInstance.Child.MemberInstance).  

In the provided example, a type-safe enum is created with members representing the enum values:

```
public class EnumEmu
{
    public int ID { get; set; }
    
    public static EnumEmu One => new EnumEmu { ID = 1 };
    // ...
}
```

This issue with this implementation is that the static member ```EnumEmu.One``` is exposed as a property which returns a new instance of ```EnumEmu``` upon each reference.  This means that given the following code, ref1 and ref2 will never refer to the same instance:
```
var ref1 = EnumEmu.One;
var ref2 = EnumEmu.One;
```

I surmise this is the reason  the methods```GetHashCode()``` and ```Equals()``` were overridden.  While this implementation works, it's inefficient.  A more efficient implementation would be as follows:

```
public class EnumEmu
{
    public int ID { get; set; }
    
    public static readonly TypeSafeEnumType One = new() {Id = 1};
    // other values
    
    // no overridden GetHashCode() or Equals() methods
}
```

Here, EnumEmu.One refers to a static object on the heap directly, not a static object's property accessor method which returns a new non-static instance.

While removing the inclusion of static members from the exception output does _hide_ the underlying issue with the example code, the encountered exceptions would still be seen if such a recursive structure were present with non-static members.

The library does currently have guards against infinite recursion, but it's mechanism relies upon object equality to determine previously-visited members when serializing output.  To address the issue presented in the example, a guard which tracked object graph depth would be needed requiring an arbitrary depth and appropriate configuration for use cases not including recursion, but with object graph depth higher than the default.  That's quite the edge case and one for which I can't envision a legitimate need (i.e. infinite recursion without object-defined equality).
